src/main/java/io/swagger/codegen/languages/kotlin/KotlinClientCodegen.java

```diff
@@ -0,0 +1,127 @@
+package io.swagger.codegen.languages.kotlin;
+
+import io.swagger.codegen.CliOption;
+import io.swagger.codegen.CodegenConstants;
+import io.swagger.codegen.CodegenType;
+import io.swagger.codegen.SupportingFile;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class KotlinClientCodegen extends AbstractKotlinCodegen {
+
+    public static final String DATE_LIBRARY = "dateLibrary";
+    protected CodegenConstants.ENUM_PROPERTY_NAMING_TYPE enumPropertyNaming = CodegenConstants.ENUM_PROPERTY_NAMING_TYPE.camelCase;
+    static Logger LOGGER = LoggerFactory.getLogger(KotlinClientCodegen.class);
+
+    protected String dateLibrary = DateLibrary.JAVA8.value;
+
+    public enum DateLibrary {
+        STRING("string"),
+        THREETENBP("threetenbp"),
+        JAVA8("java8");
+
+        public final String value;
+
+        DateLibrary(String value) {
+            this.value = value;
+        }
+    }
+
+    /**
+     * Constructs an instance of `KotlinClientCodegen`.
+     */
+    public KotlinClientCodegen() {
+        super();
+
+        artifactId = "kotlin-client";
+        packageName = "io.swagger.client";
+
+        outputFolder = "generated-code" + File.separator + "kotlin-client";
+        modelTemplateFiles.put("model.mustache", ".kt");
+        apiTemplateFiles.put("api.mustache", ".kt");
+        modelDocTemplateFiles.put("model_doc.mustache", ".md");
+        apiDocTemplateFiles.put("api_doc.mustache", ".md");
+        embeddedTemplateDir = templateDir = "kotlin-client";
+        apiPackage = packageName + ".apis";
+        modelPackage = packageName + ".models";
+
+        CliOption dateLibrary = new CliOption(DATE_LIBRARY, "Option. Date library to use");
+        Map<String, String> dateOptions = new HashMap<>();
+        dateOptions.put(DateLibrary.THREETENBP.value, "Threetenbp");
+        dateOptions.put(DateLibrary.STRING.value, "String");
+        dateOptions.put(DateLibrary.JAVA8.value, "Java 8 native JSR310");
+        dateLibrary.setEnum(dateOptions);
+        cliOptions.add(dateLibrary);
+    }
+
+    public CodegenType getTag() {
+        return CodegenType.CLIENT;
+    }
+
+    public String getName() {
+        return "kotlin-client";
```

Comment:
Not really part of this PR, but this seems as good a place as any to comment.

I think it's weird that this value is still used as an option to "language" when generating code. It makes sense that this generator's name is "kotlin-client" (more sense than when I generate scala client code at work using custom templates with `-l scalatra` -- which is a server generator and not a language).

I had suggested in an issue in the swagger-codegen repo that there are multiple levels of generation that I think should be addressed:

* language (e.g. Kotlin)
* generator type (e.g. Client)
* framework (e.g. Default, OkHttp, Retrofit)
* Options: language version, framework version, etc.

To support a lot of this cleanly, you'd want shared templates that can be reused across framework options and across language/framework versions.

Reply:
Great comment, @jimschubert. This might help with another issue we're looking to solve with regard to versioning. @HugoMario can you track that please?

Reply:
I agree regarding duplication. All of the model templates for kotlin-client and kotlin-server are currently duplicated. These should be shared. I was going to try and refactor this but the only way I could think of getting it to work at the moment was to make the client generator a library option... so "-l kotlin --library=client" and "-l kotlin --library=ktor-server". This seemed a little weird so I opted not to pursue it in this PR